DSC540-T301<br>
Week 1 Assignment<br>
Jeremy Barton

# Week 1 Assignment

In [1]:
# Imports and Preprocessing
import pandas as pd
import numpy as np

1. Load the dataset as a Pandas data frame.

pd.read_csv() loads in a csv. attaching it to an object creates a new dataframe from it

In [2]:
video_games_df = pd.read_csv('data/Video_Games_Sales_as_at_22_Dec_2016.csv')

2. Display the first ten rows of data.

`head` is used to pull the first three rows of data

In [3]:
video_games_df.head(3)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
0,Wii Sports,Wii,2006.0,Sports,Nintendo,41.36,28.96,3.77,8.45,82.53,76.0,51.0,8,322.0,Nintendo,E
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,NaN,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,Nintendo,15.68,12.76,3.79,3.29,35.52,82.0,73.0,8.3,709.0,Nintendo,E


3. Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?

`shape` is used to show the total rows and columns of a dataframe

In [4]:
video_games_df.shape

(16719, 16)

There are 16,719 rows and 16 columns. This means there are 16,719 records for this dataset and 16 fields.

4. Find the top five games by critic score

Using the `sort_values()` function to show top five games. 

In [5]:
video_games_df.sort_values(by=['Critic_Score']).head(5)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
13690,Ride to Hell,PS3,2013.0,Racing,Deep Silver,0.02,0.01,0.0,0.01,0.04,13.0,4.0,1,218.0,"Eutechnyx, Deep Silver",M
10663,Leisure Suit Larry: Box Office Bust,PS3,2009.0,Adventure,Codemasters,0.06,0.03,0.0,0.01,0.10,17.0,11.0,1.7,37.0,Team 17,M
12132,Anubis II,Wii,2007.0,Platform,Metro 3D,0.06,0.00,0.0,0.00,0.07,19.0,5.0,2.7,28.0,Data Design Interactive,E
12823,Ride to Hell,X360,2013.0,Racing,Deep Silver,0.04,0.01,0.0,0.01,0.06,19.0,14.0,1.3,433.0,"Eutechnyx, Deep Silver",M
15345,Balls of Fury,Wii,2007.0,Sports,Zoo Digital Publishing,0.02,0.00,0.0,0.00,0.02,19.0,6.0,2.8,25.0,Black Lantern Studios,E10+


5. Find the number of video games in the data frame in each genre.

`value_counts` will take a list of columns to sort based on the order of column names in the list. So this function sorts the values alphabetically.

In [6]:
video_games_df['Genre'].value_counts()

Genre
Action          3370
Sports          2348
Misc            1750
Role-Playing    1500
Shooter         1323
Adventure       1303
Racing          1249
Platform         888
Simulation       874
Fighting         849
Strategy         683
Puzzle           580
Name: count, dtype: int64

6. Find the first five games in the data frame on the SNES platform.

This time we aren't using a new function. Instead, the `==` operator filters the Platform on SNES to return only those rows with "SNES" as the Platform.

In [7]:
video_games_df[video_games_df['Platform'] == 'SNES'].head(5)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
18,Super Mario World,SNES,1990.0,Platform,Nintendo,12.78,3.75,3.54,0.55,20.61,NaN,NaN,NaN,NaN,NaN,NaN
56,Super Mario All-Stars,SNES,1993.0,Platform,Nintendo,5.99,2.15,2.12,0.29,10.55,NaN,NaN,NaN,NaN,NaN,NaN
71,Donkey Kong Country,SNES,1994.0,Platform,Nintendo,4.36,1.71,3.00,0.23,9.30,NaN,NaN,NaN,NaN,NaN,NaN
76,Super Mario Kart,SNES,1992.0,Racing,Nintendo,3.54,1.24,3.81,0.18,8.76,NaN,NaN,NaN,NaN,NaN,NaN
137,Street Fighter II: The World Warrior,SNES,1992.0,Fighting,Capcom,2.47,0.83,2.87,0.12,6.30,NaN,NaN,NaN,NaN,NaN,NaN


7. Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.

In this exercise we finally get to use the `groupby` function. Combined with the `sum` function, total global sales by publisher is pulled. This is limited to 5 rows with `.head()`.

In [8]:
video_games_df.groupby('Publisher')['Global_Sales'].sum().head(5)

Publisher
10TACLE Studios                  0.11
1C Company                       0.10
20th Century Fox Video Games     1.94
2D Boy                           0.04
3DO                             10.12
Name: Global_Sales, dtype: float64

8. Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.

In this solution the sum of North American sales is divided by the sum of Global Sales. A new column is created from this to show 49%.

In [17]:
video_games_df['NM_perc_global'] = video_games_df['NA_Sales'].sum() / video_games_df['Global_Sales'].sum()

Print the results.

In [16]:
print(f"{video_games_df['NM_perc_global'].iloc[0]:.2f}% of global sales were from North America.")

0.49% of global sales were from North America.


9. Find the number NaN entries (missing data values) in each column.

Using `isna()` returns a binary value (True of False) on whether that row of the column is an NaN value. By summing it, we can count the number of NaN values in each column.

In [10]:
video_games_df.isna().sum()

Name                  2
Platform              0
Year_of_Release     269
Genre                 2
Publisher            54
NA_Sales              0
EU_Sales              0
JP_Sales              0
Other_Sales           0
Global_Sales          0
Critic_Score       8582
Critic_Count       8582
User_Score         6704
User_Count         9129
Developer          6623
Rating             6769
NM_perc_global        0
dtype: int64

10. Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.

Knowing that it'll return an error we premptively replace non-numeric strings with NaNs.

In [14]:
video_games_df['User_Score'] = pd.to_numeric(video_games_df['User_Score'], errors='coerce')

The above ststment uses the `to_numeric()` function to convert all non-numeric rows in this column to NaN. This function is used to simply select all rows that are numeric.

The part `errors=coerce` is what actually changes the non-numeric values by forcing (coercing) any value that can't be converted to a number into NaN instead of throwing an error

In [19]:
video_games_df['User_Score']

0        8.0
1        NaN
2        8.3
3        8.0
4        NaN
        ... 
16714    NaN
16715    NaN
16716    NaN
16717    NaN
16718    NaN
Name: User_Score, Length: 16719, dtype: float64

In the below statment, NaN values are found using the `fillna()` function and populated with the median of all scores by using `median()`.

In [20]:
video_games_df['User_Score'] = video_games_df['User_Score'].fillna(video_games_df['User_Score'].median())

Quick check:

In [21]:
video_games_df['User_Score'].isna().sum()

np.int64(0)

Very nice. 👌